### DISCUSSION TITLE
Don't indent <head> & <body>

### DISCUSSION THREAD

**Discussion by Potherca** [prettier/prettier#14761] [Q&A] [answered]
My problem is similar to that described in issues #12348, #14565, #14698: I would like to prevent the `<head>` and `<body>` tags from being indented.

I also fully agree with Prettier to **not** add such a thing as a config option!

With other formatters[^1], it is not uncommon to create custom code (as a plugin, or other form of extension) to achieve this.

So, my question is:

What is the common solution for tweaking formats like this, which can not be set through config?

Any pointers to more detailed explanations (docs, blog post, whatever) would be much appreciated.

[^1]: For instance [Remark](https://remark.js.org/)

**Comment by tamish-max** [comment]
Not via config.

Prettier intentionally keeps HTML formatting options very small, and indentation of `<head>` / `<body>` isn’t one of the supported toggles. So the normal answers are:

- accept Prettier’s output,
- exclude those files/sections from Prettier,
- or build a custom plugin/printer if you really need a different HTML layout.

But there isn’t a built-in option for “don’t indent `<head>` and `<body>`”.

**Comment by Poliklot** [comment, marked as answer]
Hi, @Potherca!

I think this is exactly the kind of case where a Prettier plugin is the right escape hatch.

I agree that this probably should not become a new core Prettier option. But if a project wants a specific formatting rule like “do not indent `<head>` and `<body>` under `<html>`”, then the practical solution is to move that decision into a custom printer.

This is basically what I did in [`@poliklot/prettier-plugin-handlebars`](https://github.com/Poliklot/prettier-plugin-handlebars).

It is a Prettier plugin for HTML-heavy classic Handlebars templates. It registers its own parser and printer:

```ts
export const parsers = {
  handlebars: {
    parse,
    astFormat: "handlebars-ast",
    locStart,
    locEnd,
  },
};

export const printers = {
  "handlebars-ast": printer,
};
```

The important part is that the printer owns the formatting decisions. So things like element indentation, attribute formatting, blank lines, raw text handling, `script` / `style`, `prettier-ignore`, etc. are not controlled by extra Prettier core options — they are implemented in the plugin printer.

For your exact case, I would solve it in the same way: write a small Prettier plugin / custom printer rule that treats root-level `<html>`, `<head>`, and `<body>` the way your project wants.

You could also try using my plugin for HTML-like files by explicitly mapping those files to its parser, but keep in mind that this is not a drop-in replacement for Prettier’s built-in HTML formatter. It may change more than just `<head>` / `<body>` indentation. Personally, I like its formatting for HTML-heavy templates, but if the output does not match your project style, the better option is to use it as a reference and make a much smaller plugin that only solves this specific indentation rule.

I don’t think there are many other good options that fit Prettier’s philosophy:

- adding a core option: unlikely / against Prettier’s option philosophy;
- `prettier-ignore`: okay for one-off cases, not great project-wide;
- post-processing formatted HTML: possible, but more fragile;
- plugin / custom printer: the cleanest approach for project-specific formatting rules.

For more examples, the official Prettier plugins page has a list of community plugins:

https://prettier.io/docs/plugins#community-plugins

Hope this helps!
